ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -211,9 +228,58 @@ void VnodeCache_InvalidateCache(PerfTracer* _Nonnull perfTracer)
     RWLock_ReleaseExclusive(s_entriesLock);
 }
 
+IOReturn VnodeCache_ExportHealthData(IOExternalMethodArguments* _Nonnull arguments)
+{
+    PrjFSHealthData healthData =
+    {
+        .cacheCapacity = s_entriesCapacity,
+        .cacheEntries = s_cacheStats.cacheEntries, // cacheEntries is reset to 0 when VnodeCache_InvalidateCache is called
+        .invalidateEntireCacheCount = atomic_exchange(&s_cacheStats.healthStats[VnodeCacheHealthStat_InvalidateEntireCacheCount], 0ULL),
+        .totalCacheLookups = atomic_exchange(&s_cacheStats.healthStats[VnodeCacheHealthStat_TotalCacheLookups], 0ULL),
+        .totalLookupCollisions = atomic_exchange(&s_cacheStats.healthStats[VnodeCacheHealthStat_TotalLookupCollisions], 0ULL),
+        .totalFindRootForVnodeHits = atomic_exchange(&s_cacheStats.healthStats[VnodeCacheHealthStat_InvalidateEntireCacheCount], 0ULL),
+        .totalFindRootForVnodeMisses = atomic_exchange(&s_cacheStats.healthStats[VnodeCacheHealthStat_TotalFindRootForVnodeMisses], 0ULL),
+        .totalRefreshRootForVnode = atomic_exchange(&s_cacheStats.healthStats[VnodeCacheHealthStat_TotalRefreshRootForVnode], 0ULL),
+        .totalInvalidateVnodeRoot = atomic_exchange(&s_cacheStats.healthStats[VnodeCacheHealthStat_TotalInvalidateVnodeRoot], 0ULL),
+    };
+
+    // The buffer will come in either as a memory descriptor or direct pointer, depending on size
+    if (nullptr != arguments->structureOutputDescriptor)
+    {
+        IOMemoryDescriptor* structureOutput = arguments->structureOutputDescriptor;
+        if (sizeof(healthData) != structureOutput->getLength())
+        {
+            KextLog(
+                "VnodeCache_ExportHealthData: structure output descriptor size %llu, expected %lu\n",
+                static_cast<unsigned long long>(structureOutput->getLength()),
+                sizeof(healthData));
+            return kIOReturnBadArgument;
+        }
+
+        IOReturn result = structureOutput->prepare(kIODirectionIn);
+        if (kIOReturnSuccess == result)
+        {
+            structureOutput->writeBytes(0 /* offset */, &healthData, sizeof(healthData));
+            structureOutput->complete(kIODirectionIn);
+        }
+        
+        return result;
+    }
+
+    if (arguments->structureOutput == nullptr || arguments->structureOutputSize != sizeof(PrjFSHealthData))
+    {
+        KextLog("VnodeCache_ExportHealthData: structure output size %u, expected %lu\n", arguments->structureOutputSize, sizeof(healthData));
+        return kIOReturnBadArgument;
+    }
+
+    memcpy(arguments->structureOutput, &healthData, sizeof(healthData));
+    return kIOReturnSuccess;
+}
+
 KEXT_STATIC_INLINE void InvalidateCache_ExclusiveLocked()
 {
     memset(s_entries, 0, s_entriesCapacity * sizeof(VnodeCacheEntry));
+    atomic_exchange(&s_cacheStats.cacheEntries, 0U);
```

Comment:
Does this need to be atomic_exchange if it's under an exclusive lock?

Reply:
I'm glad you pointed out this line as we should be using `atomic_store` here rather than `atomic_exchange`.

Based on what I could find online `=` is equivalent to `atomic_store`, and so if we switch to using `atomic_store_exchange` with weak ordering it might be a bit faster that using `=`.  (I would also guess/hope the compiler would see that we're in an exclusive lock and do the right thing here).

References:
https://stackoverflow.com/questions/34915926/is-assignment-equivalent-to-load-store-for-stdatomicbool
https://en.cppreference.com/w/c/language/atomic